### PR TITLE
[website/docs]: Add Functions section with info on ctx/self

### PR DESCRIPTION
**Pull request by jmcook1186**
Creates new section for information about Fe functions. Adds three pages to the section:
 
- `index.md`: landing page containing content from original `/items/functions.md` page
- `self.md`: page explaining how `self` is used in Fe functions
- `context.md`: page explaining how `Context` is used in Fe functions

Content [here](https://github.com/ethereum/fe/issues/558) was used as the basis for the page on `Context`.

### CODE REVIEW

**Review by cburgdorf — COMMENTED** [review]
Great stuff! Just a few nitpicks!

File: docs/src/spec/functions/context.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
+# Context
+
+`Context` is used frequently in Fe smart contracts. It is used to gate access to EVM functions for reading and modifyng the blockchain.
```
[the remark this answers is not in the document]
Comment (jmcook1186) [reply docs/src/spec/functions/context.md:3]:
```suggestion
`Context` is used frequently in Fe smart contracts. It is used to gate access to EVM features for reading and modifyng the blockchain.
```

File: docs/src/spec/functions/context.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+To make Fe maximally explicit and as easy as possible to audit, these functions are gated behind a `Context` object. This is passed as an argument to functions, making it clear whether a function interacts with EVM features from the function signature alone. 
+
+For example, the following function looks pure from its signature (i.e. it is not expected to alter any blockchain data) but in reality it does modify the blockchain (by emitting a log).
+
+```rust
+struct SomeEvent{
+}
+
+pub fn looks_pure_but_isnt() {
```
[the remark this answers is not in the document]
Comment (jmcook1186) [reply docs/src/spec/functions/context.md:17]:
```suggestion
pub fn looks_pure_but_isnt() {
    // COMPILE ERROR
```

File: docs/src/spec/functions/context.md [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+}
+```
+
+Using `Context` to control access to EVM functions such as `emit` solves this problem by requiring an instance of `Context` to be passed explicitly to the function, making it clear from the function signature that the function executes some blockchain interaction. The function above, rewritten using `Context`, looks as follows:
+
+```fe
+struct SomeEvent {
+}
+
+pub fn uses_context(mut ctx: Context) {
```
Comment (jmcook1186) [inline docs/src/spec/functions/context.md:28]:
```suggestion
pub fn uses_context(ctx: Context) {
```

File: docs/src/spec/functions/context.md [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    ctx.emit(SomeEvent())
```
Comment (jmcook1186) [inline docs/src/spec/functions/context.md:29]:
```suggestion
    return ctx.block_number()
```

File: docs/src/spec/functions/context.md [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+pub fn looks_pure_but_isnt() {
+  emit(SomeEvent())
+}
+```
+
+Using `Context` to control access to EVM functions such as `emit` solves this problem by requiring an instance of `Context` to be passed explicitly to the function, making it clear from the function signature that the function executes some blockchain interaction. The function above, rewritten using `Context`, looks as follows:
+
+```fe
+struct SomeEvent {
```
Comment (jmcook1186) [inline docs/src/spec/functions/context.md:25]:
```suggestion
```

File: docs/src/spec/functions/context.md [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+}
```
Comment (jmcook1186) [inline docs/src/spec/functions/context.md:26]:
```suggestion
```

File: docs/src/spec/functions/context.md [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
```
Comment (jmcook1186) [inline docs/src/spec/functions/context.md:27]:
```suggestion
```

File: docs/src/spec/functions/context.md [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+fn retrieves_blocknumber(ctx: Context) -> u256 {
+  return ctx.block_number()
+}
+```
+
+## Context mutability
+
+All functionality that modifies the blockchain such as creating logs or contracts or transferring ether would require a mutable `Context` reference whereas read-only access such as `ctx.blocknumber()` does not need require a mutable reference to the context. To pass a mutable `Context` object, prepend the object name with `mut` in the function definition, e.g.:
```
Comment (jmcook1186) [inline docs/src/spec/functions/context.md:61]:
```suggestion
All functionality that modifies the blockchain such as creating logs or contracts or transferring ether would require a mutable `Context` reference whereas read-only access such as `ctx.block_number()` does not need require a mutable reference to the context. To pass a mutable `Context` object, prepend the object name with `mut` in the function definition, e.g.:
```

File: docs/src/spec/items/functions/context.md [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+For example, the following function looks pure from its signature (i.e. it is not expected to alter any blockchain data) but in reality it does modify the blockchain (by emitting a log).
+
+```rust
+struct SomeEvent{
+}
+
+pub fn looks_pure_but_isnt() {
+  // COMPILE ERROR
+  emit(SomeEvent())
```
Comment (cburgdorf) [inline docs/src/spec/items/functions/context.md:19]: >function above, rewritten using `Context`, looks as follows:

There's now a mismatch here.

Maybe change the above example to:

```
pub fn looks_pure_but_isnt() {
  // COMPILE ERROR
  block_number()
}
```

**Pull request merged**